Entrust cagw certificate

sapnajainEntrust opened the pull request:
##### SUMMARY
I am trying to add a new module for the crypto collection i.e. entrust_cagw_certificate and the purpose of the module is to be able to request private/public PKI certs from CAs like Entrust Security Manager, Entrust PKIaaS, Entrust SSL CA, and third party CAs like Microsoft.
This is how it is different from the existing Entrust plugin ecs_certificate within the community crypto collection -

Unlike ECS, this plugin can issue certs from both public and private Certificate Authorities supported by CAGW.
Entrust CA Gateway(CAGW) provides a REST based interface that talks to different Certificate Authorities as mentioned above and this new module is supposed to interact with CAGW over API interface.
I am at a point where I have been able to code complete the module, test with different CAs and even have my integration test suite ready.

##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
entrust_cagw_certificate

##### ADDITIONAL INFORMATION
- name: Request a certificate from Entrust Certificate Authority via Entrust CAGW with bare minimum parameters
   entrust_cagw_certificate:
      path: '{{ output_cert_path }}'
      csr: '{{ csr_path }}'
      host: '{{ entrust_host }}'
      port: '{{ entrust_port }}'
      certificate_authority_id: '{{ ca_id }}'
      certificate_profile_id: '{{ cert_profile_id }}'
      cagw_api_client_cert_path: '{{ entrust_cagw_api_cert }}'
      cagw_api_client_cert_key_path: '{{ entrust_cagw_api_cert_key }}'
      cagw_api_specification_path: '{{ cagw_api_specification_path }}'
      request_type: '{{ request_type }}'
      enrollment_format: '{{ enrollment_format }}'
      connector_name: SM
      force: '{{ force }}'
      validate_certs: '{{ validate_certs }}'

```

felixfontein commented on plugins/module_utils/entrust_cagw/api.py:
```diff
@@ -0,0 +1,347 @@
+# their own license to the complete work.
+#
+# Copyright (c), Entrust Datacard Corporation, 2020
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```
You also need to add
```suggestion
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
# SPDX-License-Identifier: BSD-2-Clause
```
(See the failing extra sanity tests, and the failing reuse test.)

sapnajainEntrust replied:
I changed the license to GPL-3.0 and added below too
 # SPDX-License-Identifier: GPL-3.0-or-later

felixfontein commented on plugins/modules/entrust_cagw_certificate.py:
```diff
@@ -0,0 +1,860 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c), Entrust Datacard Corporation, 2019
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Same here:
```suggestion
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later
```

sapnajainEntrust replied:
done

felixfontein commented on tests/integration/targets/entrust_cagw_certificate/aliases:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) Ansible Project
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
```
This marks the test as "won't run in CI":
```suggestion

unsupported

```

felixfontein commented on plugins/module_utils/entrust_cagw/api.py:
```diff
@@ -0,0 +1,347 @@
+                    cagw_api_specification_path, e.getcode())))
+        else:
+            with open(cagw_api_specification_path) as f:
+                if ".json" in cagw_api_specification_path:
```
Shouldn't you use `endswith()` instead of `in`?
```suggestion
                if cagw_api_specification_path.endswith(".json"):
```
Otherwise this will also match if `cagw_api_specification_path == '/foo.jsonpath/bla.yml'`.

felixfontein commented on plugins/module_utils/entrust_cagw/api.py:
```diff
@@ -0,0 +1,347 @@
+                if ".json" in cagw_api_specification_path:
+                    self._spec = json.load(f)
+                elif ".yml" in cagw_api_specification_path or ".yaml" in cagw_api_specification_path:
+                    self._spec = yaml.safe_load(f)
```
If none of the two branches matches, you should react as well (like raise an exception "unknown file extension" or something?).

felixfontein commented on plugins/modules/entrust_cagw_certificate.py:
```diff
@@ -0,0 +1,860 @@
+            - request type i.e. new (stands for enrollment), get (stands for get certificate), action (stands for action to be taken on the certificate)
+        type: str
+        choices: [ 'new', 'action', 'get' ]
+        required: true
```
The UX might be cleaner if you have multiple modules for these operations. Like `entrust_cagw_certificate_info` for getting a certificate, `entrust_cagw_certificate` for creating a new one, and `entrust_cagw_certificate_action` for applying an action.

sapnajainEntrust replied:
I have done this mapping vis-a-vis the certificate module within the Entrust CA Gateway where the operations i.e. new, action, and get are like crud operations on an entity i.e. certificate. I can break the module but that will be slightly mis-aligned with the CAGW API specs. Do you see this as a must have?

felixfontein commented on plugins/modules/entrust_cagw_certificate.py:
```diff
@@ -0,0 +1,860 @@
+        description:
+            - The destination path for the generated certificate as a PEM encoded cert.
+            - If there is already an Entrust certificate at this location, it will be replaced always.
+            - If enrollment_format is PKCS12 then it will have Base64 encoded PKCS12 body
```
```suggestion
            - If O(enrollment_format=PKCS12) then it will have Base64 encoded PKCS12 body.
```
1. Every paragraph should consist out of complete sentences (usually a period at the end).
2. Check out https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation and https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#semantic-markup-within-module-documentation.

felixfontein commented on plugins/modules/entrust_cagw_certificate.py:
```diff
@@ -0,0 +1,860 @@
+
+    request_type:
+        description:
+            - request type i.e. new (stands for enrollment), get (stands for get certificate), action (stands for action to be taken on the certificate)
```
Please avoid Latin abbreviations (https://docs.ansible.com/ansible/devel/dev_guide/style_guide/spelling_word_choice.html#i-e).

sapnajainEntrust replied:
done

felixfontein commented on plugins/modules/entrust_cagw_certificate.py:
```diff
@@ -0,0 +1,860 @@
+
+    host:
+        description:
+            - Host or ip address for Entrust CAGW
```
```suggestion
            - Host or IP address for Entrust CAGW.
```
Acronyms should always be upper-case (https://docs.ansible.com/ansible/devel/dev_guide/style_guide/spelling_word_choice.html#acronyms).

sapnajainEntrust replied:
@felixfontein Thanks a lot for your comments. 
Yeah I need to fix the failed sanity tests and other documentation related issues which I will take up as soon as possible.

sapnajainEntrust commented on plugins/module_utils/entrust_cagw/api.py:
```diff
@@ -0,0 +1,347 @@
+# This particular file snippet, and this file snippet only, is licensed under the
+# Modified BSD License. Modules you write using this snippet, which is embedded
+# dynamically by Ansible, still belong to the author of the module, and may assign
+# their own license to the complete work.
```
No, I used BSD-2 because I coded this module it was a part of Ansible main repository only.
I will change the license to GPLv3.

sapnajainEntrust replied:
I changed the license to GPLv3

sapnajainEntrust commented:
@felixfontein I incorporated all of your comments and fixed other indentation related issues.
As suggested by you, I added "unsupported" flag in aliases file but looks like still it is failing as shown below in sanity tests(although aliases file is in tests/integration). Do I need to do something else too to avoid these tests in CI.
00:27 FATAL: The 1 sanity test(s) listed below (out of 34) failed. See error output above for details.
00:27 pylint
00:27 Run command with stdout: docker exec -i ansible-test-controller-Hs8VreZW sh -c 'tar cf - -C /root/ansible_collections/community/crypto/tests --exclude .tmp output | gzip'
00:27 Run command with stdin: tar oxzf - -C /__w/5/ansible_collections/community/crypto/tests
00:27 FATAL: Command "docker exec ansible-test-controller-Hs8VreZW /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/crypto LC_ALL=en_US.UTF-8 /usr/bin/python3.12 /root/ansible/bin/ansible-test sanity -v --containers '{}' --junit --coverage-check --allow-disabled --truncate 0 --color yes --host-path tests/output/.tmp/host-7yqk956q --metadata tests/output/.tmp/metadata-stpya6j0.json --require plugins/module_utils/entrust_cagw/api.py --require plugins/modules/entrust_cagw_certificate.py --require tests/integration/targets/entrust_cagw_certificate/aliases --require tests/integration/targets/entrust_cagw_certificate/defaults/main.yml --require tests/integration/targets/entrust_cagw_certificate/meta/main.yml --require tests/integration/targets/entrust_cagw_certificate/tasks/main.yml --require tests/integration/targets/entrust_cagw_certificate/vars/main.yml" returned exit status 1. 
Thanks a lot in advance.

sapnajainEntrust commented:
> @felixfontein I incorporated all of your comments and fixed other indentation related issues. As suggested by you, I added "unsupported" flag in aliases file but looks like still it is failing as shown below in sanity tests(although aliases file is in tests/integration). Do I need to do something else too to avoid these tests in CI. 00:27 FATAL: The 1 sanity test(s) listed below (out of 34) failed. See error output above for details. 00:27 pylint 00:27 Run command with stdout: docker exec -i ansible-test-controller-Hs8VreZW sh -c 'tar cf - -C /root/ansible_collections/community/crypto/tests --exclude .tmp output | gzip' 00:27 Run command with stdin: tar oxzf - -C /__w/5/ansible_collections/community/crypto/tests 00:27 FATAL: Command "docker exec ansible-test-controller-Hs8VreZW /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/crypto LC_ALL=en_US.UTF-8 /usr/bin/python3.12 /root/ansible/bin/ansible-test sanity -v --containers '{}' --junit --coverage-check --allow-disabled --truncate 0 --color yes --host-path tests/output/.tmp/host-7yqk956q --metadata tests/output/.tmp/metadata-stpya6j0.json --require plugins/module_utils/entrust_cagw/api.py --require plugins/modules/entrust_cagw_certificate.py --require tests/integration/targets/entrust_cagw_certificate/aliases --require tests/integration/targets/entrust_cagw_certificate/defaults/main.yml --require tests/integration/targets/entrust_cagw_certificate/meta/main.yml --require tests/integration/targets/entrust_cagw_certificate/tasks/main.yml --require tests/integration/targets/entrust_cagw_certificate/vars/main.yml" returned exit status 1. Thanks a lot in advance.

@felixfontein I figured it out and fixed this problem. This problem was because of unused imports. Thanks..

thedoubl3j commented:
@felixfontein hold on merging please, potential partner joining and I don't to merge this just for it to be pulled out in another PR for certification.

felixfontein commented:
I would suggest to mark PRs that should maybe not merged as a draft the next time. I've done so now.

sapnajainEntrust commented:
@thedoubl3j I talked to my colleague in Entrust who is Redhat channel partner. He shared me this link below 
 to get to know about the certification process. 
 https://connect.redhat.com/sites/default/files/2021-06/MBU%20_%20Ansible%20Certification%20Policy%20Guide.pdf
 But looks like when we add a new collection then we have to go though the certification process but in my case I am just adding a module(entrust_cagw_certificate) in existing collection i.e. community.crypto.
 Does this certification make any sense here?

felixfontein commented:
Any news on this?

sapnajainEntrust commented:
> Any news on this?

@felixfontein I am working with RedHat channel partner to push it as a collection instead of making it as community.crypto module

felixfontein commented:
@sapnajainEntrust ok. In that case, would you mind closing this PR? If this isn't going to be merged in the community.crypto collection there's no reason to keep it open.

sapnajainEntrust commented:
@felixfontein I am closing this.
